labs/b4.md

```diff
@@ -0,0 +1,269 @@
+Once the compilation is done, the resulting executables are usually stored in
+`./bin` or `./build`.
+
+GCC and clang also have compiler flags that allow certain features to be
+enabled. Usually the flags that actually matter are optimization flags.
+Depending on what you want to optimize for, either space or memory or speed,
+there is a flag for it.
+
+Now, we will make a very simple application in C that prints “Hello Penguin!”
+named `hellopenguin`. Run:
+
+    $ nano hellopenguin.c
+
+to create a file named `hellopenguin.c`, and type in the following:
+
+    #include <stdio.h>
+    int main()
+    {
+        printf("Hello Penguin!");
+        return 0;
+    }
+
+Now save and exit.
+
+We will now compile the source file that you have just written:
+
+    $ gcc hellopenguin.c -o hellopenguin
+
+What this does is take the source file `hellopenguin.c` and compile it, writing
+the executable output to a file named `hellopenguin`.
+
+# How do I package stuff?
+
+Packaging manually for Debian can frustrating, especially for first timers.
+That’s why for this class, we’ll be using a really cool Ruby gem called `fpm` which
+simplifies the task of packaging a lot. Please note that if you were going to
+package for Debian’s own repositories, this is NOT the  correct/formal way to
+go about packaging. It is merely a great way to backport or package your own
+applications extremely quickly. First, make sure `ruby` and its own package
+manager called `gem` is installed. If they aren't, run `sudo apt install
+ruby-dev ruby-ffi`.  Now run the following to install `fpm` locally:
+
+    $ gem install fpm --user
+
+Try invoking `fpm` and if it doesn’t work, add `~/.gem/ruby/2.3.0/bin` to your
+`PATH`. To do that, add this to your `.bashrc`, or just type the following into
+Bash to temporarily add it to your PATH:
+
+    $ export PATH=~/.gem/ruby/2.3.0/bin:$PATH
+
+Now we will create a very simple package using the `hellopenguin` executable
+that you made above. First, we will make a new folder named `packpenguin` and
+move into it:
+
+    $ mkdir packpenguin
+
+    $ cd packpenguin
+
+Now we will create the folder structure of where the executable will reside.
+
+In Debian, user-level packages usually reside in the folder `/usr/bin/`.
+
+    $ mkdir -p usr/bin
+
+Now move your `hellopenguin` into the `packpenguin/usr/bin/` folder.
+
+    $ cd ../ # cd into the directory where the hellpenguin executable is
+    $ mv hellopenguin packpenguin/usr/bin/
+
+Now we will create a package called `hellopenguin`. Move into the parent
+directory of the hellopenguin folder and invoke the following:
+
+    $ fpm -s dir -t deb -n hellopenguin -v 1.0~ocf1 -C packpenguin
+
+This specifies that you want to take in a directory, using the `-s` flag, and
+to output a .deb package using the `-t` flag. It takes in a directory called
+`packpenguin`, using the `-C` flag, and output a .deb file named hellopenguin,
+using `-n`, with a version number of `1.0~ocf1`, using the `-v` flag.
```

Comment:
Explain why the version is formatted as such

Reply:
I think that is unnecessary detail.